### PR TITLE
[TACHYON-726] Add methods of creating TTransport

### PR DESCRIPTION
**Pull request by dongche**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-726

This patch adds a method in AuthenticationFactory for creating server side TTransportFactory, which is used to build a Thrift TServer.

In original code, `TachyonMaster` build a `TThreadPoolServer` with transport `TFramedTransport.Factory`. This method increase options of transport, so enhance the server.

### CODE REVIEW

**Comment by dongche** [comment]
@ooq , @apc999 , would you like to take a look at this PR? Thanks.

File: common/src/main/java/tachyon/security/PlainSaslServer.java [hunk 1]
```diff
@@ -100,7 +100,8 @@ public String getMechanismName() {
         throw new SaslException("AuthorizeCallback authorized failure");
       }
     } catch (AuthenticationException ae) {
-      throw new SaslException("AuthenticationProvider authenticate failed:" + ae.getMessage(), ae);
+      throw new SaslException("AuthenticationProvider authenticate failed: " + ae.getMessage(),
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: What's the change here? We should keep one line if within the limit.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 2]
```diff
@@ -85,18 +90,16 @@ public static AuthType getValidAuthType(String authTypeStr) {
...
     mTachyonConf = tachyonConf;
-    // TODO: change the default value from NOSASL to SIMPLE, after feature is stable.
-    mAuthTypeStr = tachyonConf.get(Constants.TACHYON_SECURITY_AUTHENTICATION,
-        AuthType.NOSASL.getAuthName());
+    mAuthType = getAuthTypeFromConf(tachyonConf);
   }
 
-  String getAuthTypeStr() {
-    return mAuthTypeStr;
+  AuthType getAuthType() {
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Given you are changing the method here, add "public"?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 3]
```diff
@@ -22,49 +41,218 @@
...
+      public void run() {
+        mServer.serve();
+      }
+    });
+
+    serverThread.start();
+
+    // ensure server is running
+    while (!mServer.isServing() && serverThread.isAlive()) {
+      Thread.sleep(50);
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Break if server doesn't start serving for some time, so to prevent infinite waiting on test results?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 4]
```diff
@@ -22,49 +41,218 @@
...
+    }
   }
 
-  // TODO: add more tests for methods of AuthenticationFactory
+  // FIXME: This client side Callback Handler is on-going in TACHYON-621.
+  // This code is temporarily used for test and should be deleted after TACHYON-621 merged.
+  /**
+   * A client side callback to put application provided username/pwd into SASL transport.
+   */
+  private static class PlainCallbackHandler implements CallbackHandler {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Maybe add "client" in the name to differentiate from the server counterpart.

**Comment by ooq** [comment]
@dongche Thanks for the PR, it already looks quite good to me. I left some comments. In addition, can we add another test for `customAuthenticationTest` that allows only one specific username:password pair? I think such stricter check will be useful. You may have to add another provider in `PlainServerCallbackHandlerTest`.

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 1 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: A blank is added to make the message format looks better. 
Yes, it should be in one line. Updated.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 2 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Done

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 3 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 4 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done

**Comment by dongche** [comment]
@ooq Thanks for your review! A patch is updated to address comments and change the `customAuthenticationTest` to check strictly on a username:password pair.

Could you help to review it again? Thanks.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 5]
```diff
@@ -111,5 +114,27 @@ public static AuthType getAuthTypeFromConf(TachyonConf conf) {
         AuthType.NOSASL.getAuthName()));
   }
 
-  // TODO: add methods of getting different Thrift class in follow-up PR.
+  /**
+   * For server side, this method return a TTransportFactory based on the auth type. It is used as
+   * one argument to build a Thrift TServer.
+   * @return a corresponding TTransportFactory
+   * @throws SaslException if building a TransportFactory fails
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: It also throws `UnsupportedOperationException`. (I don't remember if `UnsupportedOperationException` is also `SaslException`)

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 6]
```diff
@@ -22,49 +42,235 @@
...
+      Constants.DEFAULT_MASTER_PORT);
+
   @Rule
   public ExpectedException mThrown = ExpectedException.none();
 
   @Test
   public void authenticationFactoryConstructorTest() {
-    TachyonConf tachyonConf = new TachyonConf();
-    tachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "SIMPLE");
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "SIMPLE");
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Let's try every single authType just like `#getAuthTypeFromConfTest`?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 7]
```diff
@@ -22,49 +42,235 @@
...
 
-    Assert.assertEquals(AuthenticationFactory.AuthType.SIMPLE.getAuthName(),
-        authenticationFactory.getAuthTypeStr());
+    Assert.assertEquals(AuthenticationFactory.AuthType.SIMPLE, authenticationFactory.getAuthType());
   }
 
   @Test
   public void getAuthTypeFromConfTest() {
-    TachyonConf tachyonConf = new TachyonConf();
     AuthenticationFactory.AuthType authType;
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Let's just try every single authType?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 8]
```diff
@@ -22,49 +42,235 @@
...
+    mServer.stop();
+  }
+
+  /**
+   * In SIMPLE mode, the TTransport mechanism is PLAIN. When server authenticate the connected
+   * client user, it use {@link tachyon.security.authentication.SimpleAuthenticationProviderImpl}.
+   */
+  @Test
+  public void simpleAuthenticationTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "SIMPLE");
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: In simple mode, what if the client provides `null` as the username? What is the expected behavior?

**Comment by ooq** [comment]
@dongche Looks good! Left some comments and questions.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 5 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Oh, missed. Added in javadoc. Thanks.

`UnsupportedOperationException` -> `RuntimeException`
`SaslException` -> `IOException`

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 6 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 7 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done. Since `authenticationFactoryConstructorTest` already cover this test, I merged these 2 cases together.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 8 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Nice catch! A null username should be check by client side when build `TSaslClientTransport`, and an exception is thrown.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 5 above)
Comment (ooq) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Ah right, if it's RuntimeException, I don't think you need to add in the throws field of javadoc, explaining would be nice.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 8 above)
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Including empty user name ""?

**Comment by dongche** [comment]
@ooq Thanks for your good questions about the `null` username! Some test cases are missed and they are updated in latest patch.

Explanation about the updates:
* In `PlainSaslServer`, it does not catch the runtime exception like `IllegalStateException`, and convert it to a `SaslException`. If so, thrift lib `TSaslTransport` will not catch that exception and send back to client. (refer to https://github.com/apache/thrift/blob/master/lib/java/src/org/apache/thrift/transport/TSaslTransport.java#L316). This will cause client side suspends at `client.open()` in test like `simpleAuthenticationEmptyUserTest`.
So this update catch all kinds of exceptions and wrap with `SaslException`. Also do not see a need to differentiate them.
* Add some test to check null or empty username/password. It is a little duplicated to write them for both `SIMPLE` and `CUSTOM`, since these check is in transport layer and there is no difference for these 2 mode. But it seems not bad to fully cover the code path.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 8 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: No, empty case will be checked at server side.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java [hunk 9]
```diff
@@ -111,5 +114,28 @@ public static AuthType getAuthTypeFromConf(TachyonConf conf) {
...
+   * @throws UnsupportedOperationException if the auth type is not supported or recognized.
+   */
+  public TTransportFactory getServerTransportFactory() throws SaslException {
+    switch (mAuthType) {
+      case NOSASL:
+        return new TFramedTransport.Factory();
+      case SIMPLE:
```
Comment (apc999) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: do we intend to return `PlainServerTransport` for both `SIMPLE` and `CUSTOM`? In this case, please comment `// indent to fall through` after `case SIMPLE:`

File: common/src/main/java/tachyon/security/PlainSaslHelper.java [hunk 10]
```diff
@@ -15,13 +15,25 @@
...
+import tachyon.security.authentication.AuthenticationFactory.AuthType;
+import tachyon.security.authentication.AuthenticationProvider;
+import tachyon.security.authentication.AuthenticationProviderFactory;
+
 /**
  * Because the Java SunSASL provider doesn't support the server-side PLAIN mechanism.
  * There is a new provider {@link PlainSaslServerProvider} needed to support server-side
  * PLAIN mechanism.
  * PlainSaslHelper is used to register this provider.
+ * It also provide methods to generate PLAIN transport for server and client.
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: NIT: "provide" -> "provides"

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 11]
```diff
@@ -22,49 +43,387 @@
...
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // check case that user is empty
+    mThrown.expect(TTransportException.class);
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No authentication"
+        + " identity provided");
+    TTransport client = createClient(mTachyonConf, "", "whatever");
+    try {
```
Comment (apc999) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: why we `try` here? If we are not expecting any exception thrown, we shall not have `try` here.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 12]
```diff
@@ -22,49 +43,387 @@
...
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No password "
+        + "provided");
+    TTransport client = createClient(mTachyonConf, "tachyon", "");
+    try {
+      client.open();
+    } finally {
+      mServer.stop();
+    }
+  }
+
```
Comment (apc999) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Can we have a testcase `customAuthenticationTest()` with correct username and password?

**Comment by apc999** [comment]
Thanks @dongche  The PR look good in general, with a few comments left.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 13]
```diff
@@ -22,49 +43,387 @@
...
+    client.close();
+    mServer.stop();
+  }
+
+  /**
+   * In SIMPLE mode, the TTransport mechanism is PLAIN. When server authenticate the connected
+   * client user, it use {@link tachyon.security.authentication.SimpleAuthenticationProviderImpl}.
+   */
+  @Test
+  public void simpleAuthenticationTest() throws Exception {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: I think we don't want to allow username to be empty string for simple authentication. Is it checked on the server side already?

Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Just saw it's added. Nevermind then. Thanks!

Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Personally, I think `null` check and empty string check for username and password should not be done by the authorization provider. Because authorization can be implemented by any developers and they miss the check, it's left for the server code to deal with it.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 14]
```diff
@@ -22,49 +43,387 @@
...
+      Thread.sleep(50);
+      count --;
+    }
+  }
+
+  /**
+   * This customized authentication provider is used in CUSTOM mode. It authenticate the user by
+   * verifying the specific username:password pair.
+   */
+  public static class MockAuthenticationProvider implements AuthenticationProvider {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Because MockAuthenticationProvider strictly checks username and password, let's rename it, e.g. `exactAuthenticationProvider`?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 15]
```diff
@@ -22,49 +43,387 @@
...
+      mServer.stop();
+    }
+  }
+
+  /**
+   * In CUSTOM mode, the TTransport mechanism is PLAIN. When server authenticate the connected
+   * client user, it use configured AuthenticationProvider.
+   */
+  @Test
+  public void customAuthenticationTest() throws Exception {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Let's rename the test to reflect what it does exactly, not only the class it's testing. For example, it tests strict username/password pair, so we might call something like customAuthenticationExactNamePasswordTest

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 16]
```diff
@@ -22,49 +43,387 @@
...
+    mThrown.expectMessage("PLAIN: authorization ID and password must be specified");
+    TTransport client = createClient(mTachyonConf, "tachyon", null);
   }
 
-  // TODO: add more tests for methods of AuthenticationFactory
+  /**
+   * In CUSTOM mode, if client's username is empty, an exception should be thrown in server side.
+   */
+  @Test
+  public void customAuthenticationEmptyUserTest() throws Exception {
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: In this test, is the username None or empty?

**Comment by ooq** [comment]
@dongche Left some comments, otherwise looks good!

**Comment by dongche** [comment]
@apc999 , @ooq , thanks for your review, patch updated.

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 5 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Done

File: common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java (hunk 9 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/authentication/AuthenticationFactory.java]: Done

File: common/src/main/java/tachyon/security/PlainSaslHelper.java (hunk 10 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslHelper.java]: Done

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 12 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done. The case with correct username and password is in `customAuthenticationExactNamePasswordMatchTest()`

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 13 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Agreed. The username/password passed into authentication provides should not be null/empty.

In `PlainSaslServer::evaluateResponse()` method, it first check null/empty username/password, and if passed, then execute `mHandler.handle(cbList);`, where authentication provider in the handler take effect.
So when providers start to check, the null/empty cases are already handled, and providers do not need to take care of that.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 14 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 15 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Done

**Comment by ooq** [comment]
LGTM. Thanks for the patch! @dongche

**Comment by apc999** [comment]
+1 LGTM thanks for the patch!
